File: tests/e2e_test.js
const { chromium } = require('playwright')
const BASE_URL = 'https://el-codigo-lemon.vercel.app'
const TEST_API_KEY = process.env.TEST_API_KEY || ''
const CREDS = {
  superadmin: { tipo: 'superadmin', email: 'superadmin', pass: 'Golfo' },
  manager:    { tipo: 'manager',    email: 'test_manager@elcodigo.test',   pass: 'Test1234!' },
  staff:      { tipo: 'staff',      email: 'test_staff@elcodigo.test',     pass: 'Test1234!' },
  referidor:  { tipo: 'referidor',  email: 'test_referidor@elcodigo.test', pass: 'Test1234!' },
  agencia:    { tipo: 'agencia',    email: 'test_agencia@elcodigo.test',   pass: 'Test1234!' },
}
const results = []
function log(test, status, detail) {
  detail = detail || ''
  const icon = status === 'PASS' ? '✅' : '❌'
  console.log(icon + ' [' + status + '] ' + test + (detail ? ' — ' + detail : ''))
  results.push({ test, status, detail })
}
async function goto(page, url) {
  await page.goto(url, { waitUntil: 'commit', timeout: 30000 })
  await page.waitForTimeout(3000)
}

// Login via API directa (evita Turnstile en producción)
async function loginViaApi(browser, rol) {
  const c = CREDS[rol]
  const context = await browser.newContext()
  const page = await context.newPage()

  try {
    const res = await context.request.post(BASE_URL + '/api/auth', {
      data: { tipo: c.tipo, email: c.email, password: c.pass, cfToken: 'test' },
      headers: {
        'Content-Type': 'application/json',
        'x-test-key': TEST_API_KEY
      }
    })

    if (res.status() !== 200) {
      console.log('  [API login] Status ' + res.status() + ' para ' + rol)
      return { page, ok: false }
    }

    // La cookie auth_token ya está en el context gracias al Set-Cookie de la respuesta
    await goto(page, BASE_URL + '/' + (rol === 'superadmin' ? 'superadmin' : rol))
    const ok = page.url().includes('/' + (rol === 'superadmin' ? 'superadmin' : rol))
    return { page, ok }
  } catch(e) {
    console.log('  [API login] Error: ' + e.message)
    return { page, ok: false }
  }
}

;(async function() {
  const browser = await chromium.launch({ headless: true })

  // TEST 1: Login page
  var page = await browser.newPage()
  try {
    await goto(page, BASE_URL + '/login')
    await page.screenshot({ path: 'tests/screenshots/01_login.png', fullPage: true })
    const ok = await page.locator("input[type='password']").first().isVisible({ timeout: 5000 }).catch(function() { return false })
    log('Login: página carga correctamente', ok ? 'PASS' : 'FAIL', ok ? '' : 'No password field')
  } catch(e) { log('Login: página carga correctamente', 'FAIL', e.message) }
  await page.close()

  // TEST 2: Rutas protegidas
  page = await browser.newPage()
  try {
    var fallos = []
    var rutas = ['/superadmin','/manager','/staff','/referidor','/agencia']
    for (var i=0; i<rutas.length; i++) {
      await goto(page, BASE_URL + rutas[i])
      if (!page.url().includes('/login')) fallos.push(rutas[i] + '→' + page.url())
    }
    log('Seguridad: rutas protegidas redirigen a login', fallos.length===0 ? 'PASS' : 'FAIL', fallos.join(' | '))
  } catch(e) { log('Seguridad: rutas protegidas redirigen a login', 'FAIL', e.message) }
  await page.close()

  // TEST 3: APIs sin auth
  page = await browser.newPage()
  try {
    var eps = ['/api/analytics/superadmin','/api/analytics/manager','/api/liquidaciones','/api/referidores','/api/staff','/api/admin/clientes','/api/manager/tickets']
    var fallos2 = []
    for (var j=0; j<eps.length; j++) {
      const res = await page.request.get(BASE_URL + eps[j])
      if (![401,403].includes(res.status())) fallos2.push(eps[j] + '=' + res.status())
    }
    log('Seguridad: APIs sin auth devuelven 401/403', fallos2.length===0 ? 'PASS' : 'FAIL', fallos2.join(', '))
  } catch(e) { log('Seguridad: APIs sin auth devuelven 401/403', 'FAIL', e.message) }
  await page.close()

  // TEST 4: Credenciales incorrectas
  page = await browser.newPage()
  try {
    await goto(page, BASE_URL + '/login')
    const em = page.locator("input[type='email']").first()
    const pw = page.locator("input[type='password']").first()
    if (await em.isVisible({ timeout:3000 }).catch(function(){return false})) await em.fill('fake@fake.com')
    await pw.fill('wrongpass999')
    await page.locator("button[type='submit']").first().click()
    await page.waitForTimeout(4000)
    log('Seguridad: credenciales incorrectas mantienen en login', page.url().includes('/login') ? 'PASS' : 'FAIL', 'URL: ' + page.url())
  } catch(e) { log('Seguridad: credenciales incorrectas mantienen en login', 'FAIL', e.message) }
  await page.close()

  // TEST 5: Superadmin
  var r5 = await loginViaApi(browser, 'superadmin')
  page = r5.page
  try {
    if (!r5.ok) throw new Error('Login API falló — asegúrate de configurar TEST_API_KEY en Vercel')
    await page.screenshot({ path: 'tests/screenshots/02_superadmin.png', fullPage: true })
    log('Superadmin: login correcto', 'PASS')
    var tabs = ['Métricas','Locales','Referidores','Staff','Agencias','Pagos','Clientes']
    for (var t=0; t<tabs.length; t++) {
      try {
        const btn = page.locator('button:has-text("' + tabs[t] + '")').first()
        if (await btn.isVisible({ timeout:3000 }).catch(function(){return false})) {
          await btn.click(); await page.waitForTimeout(800)
          log('Superadmin: tab ' + tabs[t], 'PASS')
        } else log('Superadmin: tab ' + tabs[t], 'FAIL', 'no visible')
      } catch(e2) { log('Superadmin: tab ' + tabs[t], 'FAIL', e2.message) }
    }
    // Filtros fecha
    try {
      await page.locator('button:has-text("Métricas")').first().click(); await page.waitForTimeout(500)
      var presets = ['Hoy','Semana','Mes','Año','Todo']
      var foundPreset = false
      for (var p=0; p<presets.length; p++) {
        const pbtn = page.locator('button:has-text("' + presets[p] + '")').first()
        if (await pbtn.isVisible({ timeout:2000 }).catch(function(){return false})) { await pbtn.click(); await page.waitForTimeout(300); foundPreset=true }
      }
      log('Superadmin: filtros de fecha', foundPreset ? 'PASS' : 'FAIL', foundPreset ? '' : 'ningún preset visible')
    } catch(e2) { log('Superadmin: filtros de fecha', 'FAIL', e2.message) }
    // Clientes búsqueda
    try {
      await page.locator('button:has-text("Clientes")').first().click(); await page.waitForTimeout(600)
      const srch = page.locator("input[placeholder*='cliente'], input[placeholder*='Buscar']").first()
      if (await srch.isVisible({ timeout:3000 }).catch(function(){return false})) {
        await srch.fill('test'); await page.locator('button:has-text("Buscar")').first().click(); await page.waitForTimeout(1500)
      }
      log('Superadmin: tab Clientes con búsqueda', 'PASS')
    } catch(e2) { log('Superadmin: tab Clientes con búsqueda', 'FAIL', e2.message) }
    // Nueva liquidación
    try {
      await page.locator('button:has-text("Pagos")').first().click(); await page.waitForTimeout(600)
      const vis = await page.locator('button:has-text("Nueva Liquidación")').first().isVisible({ timeout:3000 }).catch(function(){return false})
      log('Superadmin: botón Nueva Liquidación', vis ? 'PASS' : 'FAIL')
    } catch(e2) { log('Superadmin: botón Nueva Liquidación', 'FAIL', e2.message) }
  } catch(e) { log('Superadmin: login correcto', 'FAIL', e.message) }
  await page.context().close()

  // TEST 6: Manager
  var r6 = await loginViaApi(browser, 'manager')
  page = r6.page
  try {
    if (!r6.ok) throw new Error('Login API falló')
    await page.screenshot({ path: 'tests/screenshots/03_manager.png', fullPage: true })
    log('Manager: login correcto', 'PASS')
    await page.waitForTimeout(1000)
    try {
      const n = await page.locator('.glass-panel').count()
      log('Manager: tarjetas de stats visibles', n>0 ? 'PASS' : 'FAIL', 'panels=' + n)
    } catch(e2) { log('Manager: tarjetas de stats visibles', 'FAIL', e2.message) }
    try {
      var mpresets = ['Hoy','Semana','Mes','Año','Todo']
      var foundM = false
      for (var mp=0; mp<mpresets.length; mp++) {
        const mbtn = page.locator('button:has-text("' + mpresets[mp] + '")').first()
        if (await mbtn.isVisible({ timeout:2000 }).catch(function(){return false})) { await mbtn.click(); await page.waitForTimeout(400); foundM=true }
      }
      log('Manager: filtros de fecha', foundM ? 'PASS' : 'FAIL')
    } catch(e2) { log('Manager: filtros de fecha', 'FAIL', e2.message) }
    try {
      const vis = await page.locator('text=Clientes de Hoy').first().isVisible({ timeout:5000 }).catch(function(){return false})
      log("Manager: 'Clientes de Hoy' visible", vis ? 'PASS' : 'FAIL')
    } catch(e2) { log("Manager: 'Clientes de Hoy' visible", 'FAIL', e2.message) }
  } catch(e) { log('Manager: login correcto', 'FAIL', e.message) }
  await page.context().close()

  // TEST 7: Staff
  var r7 = await loginViaApi(browser, 'staff')
  page = r7.page
  try {
    if (!r7.ok) throw new Error('Login API falló')
    await page.screenshot({ path: 'tests/screenshots/04_staff.png', fullPage: true })
    log('Staff: login correcto', 'PASS')
    await page.waitForTimeout(1500)
    const html = await page.content()
    const hasQr = html.toLowerCase().includes('qr') || html.toLowerCase().includes('scan') || html.toLowerCase().includes('verific')
    log('Staff: interfaz de escaneo presente', hasQr ? 'PASS' : 'FAIL')
  } catch(e) { log('Staff: login correcto', 'FAIL', e.message) }
  await page.context().close()

  // TEST 8: Referidor
  var r8 = await loginViaApi(browser, 'referidor')
  page = r8.page
  try {
    if (!r8.ok) throw new Error('Login API falló')
    await page.screenshot({ path: 'tests/screenshots/05_referidor.png', fullPage: true })
    log('Referidor: login correcto', 'PASS')
    await page.waitForTimeout(1000)
    try {
      const qrTab = page.locator('button:has-text("Mi QR")').first()
      if (await qrTab.isVisible({ timeout:3000 }).catch(function(){return false})) {
        await qrTab.click(); await page.waitForTimeout(2000)
        await page.screenshot({ path: 'tests/screenshots/05b_referidor_qr.png' })
        const img = await page.locator('img[alt="QR Code"], canvas').first().isVisible({ timeout:3000 }).catch(function(){return false})
        log('Referidor: QR personal generado', img ? 'PASS' : 'FAIL')
      } else log('Referidor: QR personal generado', 'FAIL', 'tab no visible')
    } catch(e2) { log('Referidor: QR personal generado', 'FAIL', e2.message) }
  } catch(e) { log('Referidor: login correcto', 'FAIL', e.message) }
  await page.context().close()

  // TEST 9: Agencia
  var r9 = await loginViaApi(browser, 'agencia')
  page = r9.page
  try {
    if (!r9.ok) throw new Error('Login API falló')
    await page.screenshot({ path: 'tests/screenshots/06_agencia.png', fullPage: true })
    log('Agencia: login correcto', 'PASS')
    await page.waitForTimeout(1000)
    var atabs = ['Promotores','Pagos']
    for (var at=0; at<atabs.length; at++) {
      try {
        const abtn = page.locator('button:has-text("' + atabs[at] + '")').first()
        const avis = await abtn.isVisible({ timeout:3000 }).catch(function(){return false})
        if (avis) { await abtn.click(); await page.waitForTimeout(500); log('Agencia: tab ' + atabs[at], 'PASS') }
        else log('Agencia: tab ' + atabs[at], 'FAIL', 'no visible')
      } catch(e2) { log('Agencia: tab ' + atabs[at], 'FAIL', e2.message) }
    }
  } catch(e) { log('Agencia: login correcto', 'FAIL', e.message) }
  await page.context().close()

  // TEST 10: Landing
  page = await browser.newPage()
  try {
    await goto(page, BASE_URL)
    await page.screenshot({ path: 'tests/screenshots/07_landing.png', fullPage: true })
    const ok = await page.locator('text=El Código').first().isVisible({ timeout:5000 }).catch(function(){return false})
    log('Landing: página principal carga', ok ? 'PASS' : 'FAIL', ok ? '' : 'title=' + await page.title())
  } catch(e) { log('Landing: página principal carga', 'FAIL', e.message) }
  await page.close()

  await browser.close()
  console.log('\n' + '='.repeat(60))
  console.log('RESUMEN DE TESTS')
  console.log('='.repeat(60))
  const passed = results.filter(function(r){return r.status==='PASS'}).length
  const failed = results.filter(function(r){return r.status==='FAIL'}).length
  console.log('✅ Pasados: ' + passed + '/' + results.length)
  console.log('❌ Fallados: ' + failed + '/' + results.length)
  if (failed > 0) {
    console.log('\nFallos:')
    results.filter(function(r){return r.status==='FAIL'}).forEach(function(r){
      console.log('  • ' + r.test + ': ' + r.detail)
    })
  }
  console.log('='.repeat(60))
  process.exit(failed === 0 ? 0 : 1)
})()

File: tests/e2e_test.py
"""
Tests E2E para El Código - https://el-codigo-lemon.vercel.app
Credenciales de test creadas en Supabase para cada rol.
"""
from playwright.sync_api import sync_playwright, expect
import sys
import time

BASE_URL = "https://el-codigo-lemon.vercel.app"

CREDS = {
    "superadmin": {"tipo": "superadmin", "email": "superadmin", "pass": "Golfo"},
    "manager":    {"tipo": "manager",    "email": "test_manager@elcodigo.test",   "pass": "Test1234!"},
    "staff":      {"tipo": "staff",      "email": "test_staff@elcodigo.test",     "pass": "Test1234!"},
    "referidor":  {"tipo": "referidor",  "email": "test_referidor@elcodigo.test", "pass": "Test1234!"},
    "agencia":    {"tipo": "agencia",    "email": "test_agencia@elcodigo.test",   "pass": "Test1234!"},
}

results = []

def log(test, status, detail=""):
    icon = "✅" if status == "PASS" else "❌"
    print(f"{icon} [{status}] {test}" + (f" — {detail}" if detail else ""))
    results.append({"test": test, "status": status, "detail": detail})

def login(page, rol):
    """Hace login con el rol dado, devuelve True si tuvo éxito."""
    c = CREDS[rol]
    page.goto(f"{BASE_URL}/login")
    page.wait_for_load_state("networkidle")

    # Seleccionar tipo
    try:
        sel = page.locator(f"button:has-text('{c['tipo'].capitalize()}'), [data-tipo='{c['tipo']}']").first
        if sel.is_visible():
            sel.click()
    except:
        pass

    # Rellenar campos
    email_field = page.locator("input[type='email'], input[placeholder*='mail'], input[placeholder*='correo']").first
    pass_field  = page.locator("input[type='password']").first

    if c["tipo"] == "superadmin":
        # Superadmin solo usa contraseña
        try:
            pass_field.fill(c["pass"])
        except:
            email_field.fill(c["email"])
            pass_field.fill(c["pass"])
    else:
        email_field.fill(c["email"])
        pass_field.fill(c["pass"])

    page.locator("button[type='submit'], button:has-text('Entrar'), button:has-text('Acceder')").first.click()
    page.wait_for_load_state("networkidle")
    time.sleep(2)
    return page.url != f"{BASE_URL}/login"


with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)

    # ─────────────────────────────────────────
    # TEST 1: Página de login carga correctamente
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        page.goto(f"{BASE_URL}/login")
        page.wait_for_load_state("networkidle")
        page.screenshot(path="tests/screenshots/01_login.png")
        assert "El Código" in page.title() or page.locator("input[type='password']").is_visible()
        log("Login: página carga correctamente", "PASS")
    except Exception as e:
        log("Login: página carga correctamente", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 2: Rutas protegidas redirigen a login
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        for ruta in ["/superadmin", "/manager", "/staff", "/referidor", "/agencia"]:
            page.goto(f"{BASE_URL}{ruta}")
            page.wait_for_load_state("networkidle")
            assert "/login" in page.url or page.locator("input[type='password']").is_visible(), f"{ruta} no redirige a login"
        log("Seguridad: rutas protegidas redirigen a login", "PASS")
    except Exception as e:
        log("Seguridad: rutas protegidas redirigen a login", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 3: APIs sin auth devuelven 401
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        endpoints = [
            "/api/analytics/superadmin",
            "/api/analytics/manager",
            "/api/liquidaciones",
            "/api/referidores",
            "/api/lugares",
            "/api/staff",
        ]
        for ep in endpoints:
            res = page.request.get(f"{BASE_URL}{ep}")
            assert res.status in [401, 403], f"{ep} devolvió {res.status} (esperado 401/403)"
        log("Seguridad: APIs sin auth devuelven 401/403", "PASS")
    except Exception as e:
        log("Seguridad: APIs sin auth devuelven 401/403", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 4: Login superadmin
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        ok = login(page, "superadmin")
        page.screenshot(path="tests/screenshots/02_superadmin.png")
        assert ok and "/superadmin" in page.url
        log("Superadmin: login correcto", "PASS")

        # TEST 5: Tabs del superadmin
        for tab_text in ["Métricas", "Locales", "Referidores", "Staff", "Agencias", "Pagos", "Clientes"]:
            try:
                btn = page.locator(f"button:has-text('{tab_text}')").first
                btn.click()
                page.wait_for_load_state("networkidle")
                time.sleep(0.5)
                log(f"Superadmin: tab '{tab_text}' carga", "PASS")
            except Exception as e:
                log(f"Superadmin: tab '{tab_text}' carga", "FAIL", str(e))

        # TEST 6: Filtros de fecha en analytics
        try:
            page.locator("button:has-text('Métricas')").first.click()
            page.wait_for_load_state("networkidle")
            for preset in ["Hoy", "Semana", "Mes", "Año", "Todo"]:
                btn = page.locator(f"button:has-text('{preset}')").first
                if btn.is_visible():
                    btn.click()
                    time.sleep(0.3)
            log("Superadmin: filtros de fecha analytics", "PASS")
        except Exception as e:
            log("Superadmin: filtros de fecha analytics", "FAIL", str(e))

        # TEST 7: Tab clientes con búsqueda
        try:
            page.locator("button:has-text('Clientes')").first.click()
            page.wait_for_load_state("networkidle")
            search = page.locator("input[placeholder*='cliente']").first
            if search.is_visible():
                search.fill("a")
                page.locator("button:has-text('Buscar')").first.click()
                time.sleep(1)
            log("Superadmin: búsqueda de clientes", "PASS")
        except Exception as e:
            log("Superadmin: búsqueda de clientes", "FAIL", str(e))

        # TEST 8: Exportar Excel
        try:
            page.locator("button:has-text('Exportar Excel')").first.click()
            time.sleep(1)
            log("Superadmin: botón exportar Excel visible", "PASS")
        except Exception as e:
            log("Superadmin: botón exportar Excel visible", "FAIL", str(e))

    except Exception as e:
        log("Superadmin: login correcto", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 9: Login manager
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        ok = login(page, "manager")
        page.screenshot(path="tests/screenshots/03_manager.png")
        assert ok and "/manager" in page.url
        log("Manager: login correcto", "PASS")

        page.wait_for_load_state("networkidle")
        time.sleep(1)

        # TEST 10: Tarjetas de stats visibles
        try:
            assert page.locator("text=Volumen").first.is_visible() or page.locator("text=Comisión").first.is_visible()
            log("Manager: tarjetas de stats visibles", "PASS")
        except Exception as e:
            log("Manager: tarjetas de stats visibles", "FAIL", str(e))

        # TEST 11: Filtros de fecha
        try:
            for preset in ["Hoy", "Semana", "Mes"]:
                btn = page.locator(f"button:has-text('{preset}')").first
                if btn.is_visible():
                    btn.click()
                    time.sleep(0.3)
            log("Manager: filtros de fecha", "PASS")
        except Exception as e:
            log("Manager: filtros de fecha", "FAIL", str(e))

        # TEST 12: Sección clientes de hoy visible
        try:
            assert page.locator("text=Clientes de Hoy").first.is_visible()
            log("Manager: sección 'Clientes de Hoy' visible", "PASS")
        except Exception as e:
            log("Manager: sección 'Clientes de Hoy' visible", "FAIL", str(e))

    except Exception as e:
        log("Manager: login correcto", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 13: Login staff
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        ok = login(page, "staff")
        page.screenshot(path="tests/screenshots/04_staff.png")
        assert ok and "/staff" in page.url
        log("Staff: login correcto", "PASS")

        # TEST 14: Escáner QR visible
        try:
            time.sleep(1)
            content = page.content()
            assert "scanner" in content.lower() or "qr" in content.lower() or page.locator("video, canvas, #qr-reader").first.is_visible(timeout=3000)
            log("Staff: escáner QR presente", "PASS")
        except Exception as e:
            log("Staff: escáner QR presente", "FAIL", str(e))

    except Exception as e:
        log("Staff: login correcto", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 15: Login referidor
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        ok = login(page, "referidor")
        page.screenshot(path="tests/screenshots/05_referidor.png")
        assert ok and "/referidor" in page.url
        log("Referidor: login correcto", "PASS")

        # TEST 16: Dashboard con stats
        try:
            time.sleep(1)
            assert page.locator("text=Rendimiento, text=Comisión, text=QR").count() > 0 or page.locator("text=Tickets").first.is_visible()
            log("Referidor: dashboard con stats visible", "PASS")
        except Exception as e:
            log("Referidor: dashboard con stats visible", "FAIL", str(e))

        # TEST 17: Tab QR
        try:
            page.locator("button:has-text('Mi QR')").first.click()
            time.sleep(1)
            page.screenshot(path="tests/screenshots/05b_referidor_qr.png")
            assert page.locator("img[alt='QR Code']").is_visible(timeout=5000) or page.locator("canvas").first.is_visible(timeout=3000)
            log("Referidor: tab QR muestra código", "PASS")
        except Exception as e:
            log("Referidor: tab QR muestra código", "FAIL", str(e))

    except Exception as e:
        log("Referidor: login correcto", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 18: Login agencia
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        ok = login(page, "agencia")
        page.screenshot(path="tests/screenshots/06_agencia.png")
        assert ok and "/agencia" in page.url
        log("Agencia: login correcto", "PASS")

        # TEST 19: Tabs de agencia
        try:
            time.sleep(1)
            for tab in ["Promotores", "Pagos"]:
                btn = page.locator(f"button:has-text('{tab}')").first
                if btn.is_visible():
                    btn.click()
                    time.sleep(0.3)
            log("Agencia: tabs Promotores y Pagos visibles", "PASS")
        except Exception as e:
            log("Agencia: tabs Promotores y Pagos visibles", "FAIL", str(e))

    except Exception as e:
        log("Agencia: login correcto", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 20: Landing page
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        page.goto(BASE_URL)
        page.wait_for_load_state("networkidle")
        page.screenshot(path="tests/screenshots/07_landing.png")
        assert page.locator("text=El Código").first.is_visible()
        log("Landing: página principal carga", "PASS")
    except Exception as e:
        log("Landing: página principal carga", "FAIL", str(e))
    page.close()

    # ─────────────────────────────────────────
    # TEST 21: Credenciales incorrectas dan error
    # ─────────────────────────────────────────
    page = browser.new_page()
    try:
        page.goto(f"{BASE_URL}/login")
        page.wait_for_load_state("networkidle")
        email_field = page.locator("input[type='email']").first
        pass_field  = page.locator("input[type='password']").first
        email_field.fill("noexiste@test.com")
        pass_field.fill("wrongpassword")
        page.locator("button[type='submit'], button:has-text('Entrar')").first.click()
        time.sleep(2)
        assert "/login" in page.url
        log("Seguridad: credenciales incorrectas mantienen en login", "PASS")
    except Exception as e:
        log("Seguridad: credenciales incorrectas mantienen en login", "FAIL", str(e))
    page.close()

    browser.close()

# ─────────────────────────────────────────
# RESUMEN FINAL
# ─────────────────────────────────────────
print("\n" + "="*60)
print("RESUMEN DE TESTS")
print("="*60)
passed = sum(1 for r in results if r["status"] == "PASS")
failed = sum(1 for r in results if r["status"] == "FAIL")
total  = len(results)
print(f"✅ Pasados: {passed}/{total}")
print(f"❌ Fallados: {failed}/{total}")
if failed > 0:
    print("\nFallos:")
    for r in results:
        if r["status"] == "FAIL":
            print(f"  • {r['test']}: {r['detail']}")
print("="*60)
sys.exit(0 if failed == 0 else 1)
